rpc/src/responses/json_array.rs

```diff
@@ -1,47 +1,280 @@
-            let parsed_item_value = match nested_entry.is_array() {
-                true => unwrap_item_in_nested_json_array(nested_entry.take())?,
-                false => nested_entry.take(),
-            };
-            let converted_item: T = serde_json::from_value(parsed_item_value)?;
-            flattened_vec.push(converted_item);
+        let mut items = Vec::new();
+        for item in array {
+            let converted_item = serde_json::from_value(item.take())?;
+            items.push(converted_item);
         }
+        Ok(Self { items })
+    }
 
-        Ok(Self { flattened_vec })
+    pub fn from_json_object(json_value: &mut Value) -> Result<Self, ParseError> {
+        let converted_object = serde_json::from_value(json_value.take())?;
+        let items = vec![converted_object];
+        Ok(Self { items })
     }
 
-    pub fn into_vec(self) -> Vec<T> {
-        self.flattened_vec
+    pub fn len(&self) -> usize {
+        self.items.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
     }
 }
 
-fn unwrap_item_in_nested_json_array(mut nested_item: Value) -> Result<Value, ParseError> {
-    let nested_array = nested_item
-        .as_array_mut()
-        .ok_or_else(|| generate_none_error("invalid initial json array"))?;
-    match nested_array.last() {
-        Some(_) => Ok(nested_array.pop().unwrap()),
-        None => Ok(json!("")),
+impl<T: de::DeserializeOwned> JsonArray<JsonArray<T>> {
+    pub fn from_nested_response_str(
+        nested_json_str: &str,
+    ) -> Result<JsonArray<JsonArray<T>>, ParseError> {
+        let outer_array = try_json_str_into_json_array(nested_json_str)?;
+
+        let mut items = Vec::new();
+        for mut item in outer_array {
+            let json_item = match &item {
+                Value::Array(_) => JsonArray::from_json_array(&mut item),
+                Value::Object(_) => JsonArray::from_json_object(&mut item),
+                _ => {
+                    let detail = "initial json value is neither an object or an array".to_string();
+                    Err(ParseError::ResponseParsingError(detail))
+                }
+            }?;
+            items.push(json_item);
+        }
+        Ok(JsonArray { items })
     }
 }
 
+impl<T> iter::IntoIterator for JsonArray<T> {
+    type Item = T;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.items.into_iter()
+    }
+}
+
+impl<T: fmt::Display> fmt::Display for JsonArray<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.items.is_empty() {
+            write!(f, "")
+        } else {
+            let mut display_string = String::new();
+            for value in &self.items {
+                display_string.push_str(&format!("{}, ", value));
+            }
+            display_string.truncate(display_string.len() - 2);
+            write!(f, "{}", display_string)
```

Comment:
This is really nice, but since this is `JsonArray` shouldn't it include the outer most array?
So instead of `"{}"` in the `write!` it would be `"[{}]"`, same goes for when its empty

Reply:
That sounds more consistent with `vec` and `array` display, good point!